### DISCUSSION TITLE
YieldDay values: strange behaviour. Bug?

### DISCUSSION THREAD

**Discussion by ghost** [tbnobody/OpenDTU#630] [Q&A]
I'm testing a situation with two inverters - HM-600 and HM-1500 (entered in OpenDTU in this order).

During the day, everything works as expected, but in the last minutes before sunset, when _producing_ is already false (but _reachable_ still true) there was a point where all YieldDay values of HM-1500 suddenly dropped to zero, while the HM-600 values remained as they were. Strange! The _total_ YieldDay value decreased accordingly to the "new sum" of both, thus representing only the HM-600 value anymore.

This problem seems to be of sporadic nature; I saw it only on some days, while other days were fine. Firmware is newest (97bc964), but same behavior has also been encountered with an earlier version (b5d099f).

I'm using a RPi with a periodically cron-invoked shell script to log the complete /api/livedata/status data every minute between sunrise and sunset. So I can exactly reconstruct what happened in the last minutes of a "bad" example day.

The following has been distilled from the logged JSON data - only all the **YieldDay** values:

```
                Inverter 1 (HM-600)             Inverter 2 (HM-1500)
                -------------------     -----------------------------------
Timestamp       AC      DC0     DC1     AC      DC0     DC1     DC2     DC3     total
-------------   ---     ---     ---     ---     ---     ---     ---     ---     -----
20230224.1728   438     218     220     531     163     5       361     2       969
20230224.1729   438     218     220     531     163     5       361     2       969
20230224.1730   438     218     220     531     163     5       361     2       969
20230224.1731   438     218     220     0       0       0       0       0       438
20230224.1732   438     218     220     0       0       0       0       0       438
20230224.1733   438     218     220     0       0       0       0       0       438
20230224.1734   438     218     220     0       0       0       0       0       438
20230224.1735   438     218     220     0       0       0       0       0       438
```

The table shows that the aggregated values vanish in the 17:31 data record. Maybe important: One record before (17:30), _reachable_ of HM-1500 was false, but only this one time - perhaps a momentary disturbance of the radio link. Could this have caused the data loss??? Before and after, "reachable" was true, so OpenDTU should have had the opportunity to fetch all needed inverter data again in the remaining time. But obviously, it didn't.

So it's a pity that the web-displayed data, especially the daily summary, seemingly can't be trusted, at least after the end of the day's production timespan.

**Comment by tbnobody** [comment]
Seems to be the same "issue" as here: https://github.com/tbnobody/OpenDTU/issues/401

**Comment by ghost** [comment]
Okay, thanks for the hint!

So if I understand #401 right and try to apply your conclusion to my case, my inverter must have been _really off_ for a moment (not only a short disruption of the radio link) and therefore forgot the aggregated YieldDay values.

If this should be confirmed (I will continue to watch things), what about the following thoughts:

All YieldDay values are by definition _cumulative_ and can only grow during the respective day or remain constant, but never decrease. OpenDTU could make use of this fact to detect and correct such errors. It could compare each newly fetched YieldDay value with the previous one and always take the higher value of both instead of the newest. (Only within the same day, of course). This could make OpenDTU more robust against such kind of "glitches" and prevent wrong readings.

**Comment by tbnobody** [comment]
But as you know, memory is limited. And it would also mean that the current value has to be persistet to the flash. otherwiese the value is lost in case of a powercycle of  the esp. As mentioned in the issue, just use the yield total value. it's the most save approach.

**Comment by ghost** [comment]
Okay, I understand. But I find it not quite conclusive from a user's point of view. Please give me one last chance to plea for a fix - I will then be quiet and never insist again in this matter. (And please, don't get me wrong: I respect and highly appreciate your professional work and your decisions, so this is no criticism, but only a proposal.)

If the web UI under certain conditions shows wrong summary values, it should be called what it is: a bug. Not primarily a bug of OpenDTU itself, but a bug in its handling of a known inverter behavior which can happen anytime during regular operation.

It's a _malicious_ bug, as you can't necessarily tell from the values whether they are correct or not. To put it pointedly: all YieldDay values must be merely seen as "hints", but cannot be trusted generally.

"Just use the yield total value, it's the most save approach" - I don't understand how this can be a practical solution. It may help during automated processing of the data, but for a human's quick glance at the respective day's harvest, it doesn't help. (At least I don't know how it could.)

We are talking about two different causes with a nearly similar effect: a) inverter "glitch" (which was my focus), and, as you pointed out, b) ESP powercycling. (I agree, frequent flash writes to handle the latter are not a reasonable option due to faster wear-out.)

So for case (b) there is no simple solution. But is that a reason to dislike also one for (a), although it would be fairly easy there? (Assuming, of course, that the memory shortage you mentioned is not so dramatic that it makes the proposed solution impossible - I can't judge that).

I think from a usability point of view, case (b) is much less bad, because it's more obvious, and consequences of power interruptions are familiar to all of us, while (a) is much more insidious and harder to see through. So there are imho good reasons for a bugfix at least for this case in order to get reliable display values.

End of plea. And again, thanks for your great work!
